snippets/SharedHandle.hpp

```diff
@@ -46,45 +77,47 @@ class SharedHandleBase
   SharedHandleBase() = default;
 
   template <typename... Args>
-  SharedHandleBase( HandleType handle, Args &&... control_args ) VULKAN_HPP_NOEXCEPT
-    : m_handle( handle )
-    , m_control( std::make_shared<HeaderType>( std::forward<Args>( control_args )... ) )
+  SharedHandleBase( HandleType handle, Args &&... control_args )
+    : m_control( new ReferenceCounter<HeaderType>( std::forward<Args>( control_args )... ) ), m_handle( handle )
   {
   }
 
   SharedHandleBase( const SharedHandleBase & o ) VULKAN_HPP_NOEXCEPT
-    : m_handle( o.m_handle )
-    , m_control( o.m_control )
   {
+    o.addRef();
+    m_handle  = o.m_handle;
+    m_control = o.m_control;
   }
 
   SharedHandleBase( SharedHandleBase && o ) VULKAN_HPP_NOEXCEPT
     : m_handle( o.m_handle )
-    , m_control( std::move( o.m_control ) )
+    , m_control( o.m_control )
   {
-    o.m_handle = nullptr;
+    o.m_handle  = nullptr;
+    o.m_control = nullptr;
   }
 
-  SharedHandleBase & operator=( SharedHandleBase && o ) VULKAN_HPP_NOEXCEPT
+  SharedHandleBase & operator=( const SharedHandleBase & o ) VULKAN_HPP_NOEXCEPT
   {
-    reset();
-    m_handle   = o.m_handle;
-    m_control  = std::move( o.m_control );
-    o.m_handle = nullptr;
+    SharedHandleBase( o ).swap( *this );
```

Comment:
A `swap` on copy assignment?

Reply:
correct, because we need to remove what was there before copy assignment, so we swap the contents, new contents go to this and the old contents land to the local variable which is getting destroyed at the end of the function call, super elegant.